File: backend/app/schemas/audit.py
from __future__ import annotations

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class AuditLogEventItem(BaseModel):
    """Schema representing a single chronologically ordered audit log event with node context."""

    id: str
    node_id: Optional[str] = None
    hostname: Optional[str] = None
    device_name: Optional[str] = None
    os: Optional[str] = None
    tailnet: Optional[str] = None
    event_type: str
    event_category: str
    severity: str
    action: str
    actor: Optional[str] = None
    message: Optional[str] = None
    details: Dict[str, Any] = Field(default_factory=dict)
    ip_address: Optional[str] = None
    created_at: str

    model_config = ConfigDict(populate_by_name=True)


class AuditFeedSummary(BaseModel):
    """Breakdown summary of audit log events by severity and category."""

    total_matching: int = 0
    by_severity: Dict[str, int] = Field(default_factory=dict)
    by_category: Dict[str, int] = Field(default_factory=dict)

    model_config = ConfigDict(populate_by_name=True)


class AuditFeedResponse(BaseModel):
    """Chronological audit log feed response with filtering, search, and pagination metadata."""

    total: int
    count: int
    limit: int
    offset: int
    has_more: bool
    order: str = "desc"  # 'desc' (chronologically newest first) or 'asc' (oldest first)
    summary: AuditFeedSummary = Field(default_factory=AuditFeedSummary)
    events: List[AuditLogEventItem] = Field(default_factory=list)
    timestamp: str

    model_config = ConfigDict(populate_by_name=True)

File: frontend/src/types/audit.ts
import type { AuditEventType, AuditSeverity, EventCategory } from './enums';

export interface AuditLogEventItem {
  id: string;
  node_id: string | null;
  hostname: string | null;
  device_name: string | null;
  os: string | null;
  tailnet: string | null;
  event_type: AuditEventType | string;
  event_category: EventCategory | string;
  severity: AuditSeverity;
  action: string;
  actor: string | null;
  message: string | null;
  details: Record<string, unknown>;
  ip_address: string | null;
  created_at: string;
}

export interface AuditFeedSummary {
  total_matching: number;
  by_severity: Record<AuditSeverity | string, number>;
  by_category: Record<EventCategory | string, number>;
}

export interface AuditFeedResponse {
  total: number;
  count: number;
  limit: number;
  offset: number;
  has_more: boolean;
  order: 'desc' | 'asc';
  summary: AuditFeedSummary;
  events: AuditLogEventItem[];
  timestamp: string;
}

export interface NodeAuditFeedResponse extends AuditFeedResponse {
  error?: string;
  node_id?: string;
}

export interface AuditFeedQueryParams {
  node_id?: string;
  tailnet?: string;
  severity?: AuditSeverity | string;
  category?: EventCategory | string;
  event_type?: AuditEventType | string;
  actor?: string;
  since?: string;
  until?: string;
  q?: string;
  order?: 'desc' | 'asc';
  limit?: number;
  offset?: number;
}
